File: src/Linked_List_Cycle.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        node1, node2 = head, head
        while node1 and node2:
            node1 = node1.next
            if node2.next:
                node2 = node2.next.next
            else:
                node2 = node2.next
            if node1 and node1 == node2:
                return True
        return False
        

File: src/algorithms/cpp/Linked_List_Cycle.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *node1 = head, *node2 = head;
        while (node1 && node2) {
            node1 = node1->next;
            if (node2->next)    node2 = node2->next->next;
            else                node2 = node2->next;
            if (node1 && node1 == node2) return true;
        }
        return false;
    }
};

int main() {
    return 0;
}

